astropy/table/table.py

```diff
@@ -618,6 +623,39 @@ def __init__(self, data=None, masked=False, names=None, dtype=None,
         if self.masked not in (None, True, False):
             raise TypeError("masked property must be None, True or False")
 
+        self._set_column_attribute('unit', units)
+        self._set_column_attribute('description', descriptions)
+
+    def _set_column_attribute(self, attr, values):
+        """Set ``attr`` for columns to ``values``, which can be either a dict (keyed by column
+        name) or a dict of name: value pairs.  This is used for handling the ``units`` and
+        ``descriptions`` kwargs to ``__init__``.
+        """
+        if not values:
+            return
+
+        if isinstance(values, Row):
+            # For a Row object transform to an equivalent dict.
+            values = {name: values[name] for name in values.colnames}
+
+        if not isinstance(values, dict):
```

Comment:
Might even register `Row` as `collections.abc.Mapping`, and then do the `instance` on that (I'm a bit puzzled it is considered a `Sequence` now, since one cannot index with the column number anyway; similarly, `Row.__iter__` is weird.

But really out of scope here! See #9687